### PR TITLE
feat: support the operator make use of the config-reloader tls and basic auth with prometheus/alertmanager webConfigFile

### PR DESCRIPTION
**Pull request by dongjiang1989**
…thentication

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

ref: #6068 #5932 

Support the operator make use of the config-reloader tls and basic authentication with prometheus/alertmanager webConfigFile

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
support the operator make use of the config-reloader tls and basic auth with prometheus/alertmanager webConfigFile
```

### CODE REVIEW

**Comment by dongjiang1989** [comment]
@simonpasquier  IIUC, reuse the Prometheus/Alertmanager web config for the reloader. PTAL

**Review by simonpasquier — COMMENTED** [review]
Thanks! Got one question regarding the change in pkg/operator/defaults.go

File: pkg/operator/defaults.go [hunk 1]
```diff
@@ -31,6 +31,8 @@ const (
 	DefaultThanosBaseImage = "quay.io/thanos/thanos"
 	// DefaultThanosImage is a default image pulling address for the Thanos long-term prometheus storage collector
 	DefaultThanosImage = DefaultThanosBaseImage + ":" + DefaultThanosVersion
+	// DefaultConfigReloaderVersion is a default image tag for the prometheus-operator config-reloader
+	DefaultConfigReloaderVersion = "v0.69.0"
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/operator/defaults.go:35]: Hmmm... Be defined but not used.
Removed it

File: pkg/prometheus/agent/statefulset.go [hunk 2]
```diff
@@ -207,9 +220,11 @@ func makeStatefulSetSpec(
 		if err != nil {
 			return nil, err
 		}
+		configReloaderWebConfigFile = confArg.Value
 		promArgs = append(promArgs, confArg)
 		volumes = append(volumes, configVol...)
 		promVolumeMounts = append(promVolumeMounts, configMount...)
+		configReloaderVolumeMounts = append(configReloaderVolumeMounts, configMount...)
```
Comment (simonpasquier) [inline pkg/prometheus/agent/statefulset.go:227]: We could make an argument that adding the flag in all cases introduces a risk of breaking users not deploying the latest config-reloader tag. It would be safer to add the flag only when the web field isn't nil and we can remove the if condition in a few releases.

```suggestion
		// To avoid breaking users deploying an old version of the config-reloader image.
		// TODO: remove the if condition after v0.72.0.
		if cpf.Web != nil {
			configReloaderWebConfigFile = confArg.Value
			configReloaderVolumeMounts = append(configReloaderVolumeMounts, configMount...)
		}
```

Comment (dongjiang1989) [reply pkg/prometheus/agent/statefulset.go:227]: Thanks for your review. 

Got it. But `cpf.Web != nil` only determines whether `Prometheus's web` supports it, not whether the `config-reloader` supports it.

```
		if c.ReloaderConfig.Image.Version >= 0.69.0 {
		     // xxx
		}
```
Because config-reloader v0.69.0 already supports this feature.

Comment (dongjiang1989) [reply pkg/prometheus/agent/statefulset.go:227]: OR Add flag for Prometheus/alertmanager CRD?

Comment (simonpasquier) [reply pkg/prometheus/agent/statefulset.go:227]: My intention was only to avoid breaking users that don't specify web config for their Prometheus/Alertmanager. For others, they still need to make sure that they use a recent-enough config-reloader image.

File: pkg/prometheus/server/statefulset.go [hunk 3]
```diff
@@ -285,6 +298,13 @@ func makeStatefulSetSpec(
 		promArgs = append(promArgs, confArg)
 		volumes = append(volumes, configVol...)
 		promVolumeMounts = append(promVolumeMounts, configMount...)
+
+		// To avoid breaking users deploying an old version of the config-reloader image.
+		// TODO: remove the if condition after v0.72.0.
+		if prompkg.ConfigReloaderWebConfigFileSupported(c) {
```
Comment (dongjiang1989) [inline pkg/prometheus/server/statefulset.go:304]: Add config-reloader image version check.

File: pkg/prometheus/agent/statefulset.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/prometheus/agent/statefulset.go]: @simonpasquier Add config-reloader image version check. PTAL

**Review by simonpasquier — COMMENTED** [review]
I wouldn't try to deduce the binary version from the image tag: it won't work for folks that rebuild the images or use SHA1 refs instead of tags.
The safest approach from my POV is like I described earlier: ony pass the `--web-config-file` flag when the Alertmanager/Prometheus resource defines a non-nil `.spec.web` field.

**Comment by dongjiang1989** [comment]
> I wouldn't try to deduce the binary version from the image tag: it won't work for folks that rebuild the images or use SHA1 refs instead of tags. The safest approach from my POV is like I described earlier: ony pass the `--web-config-file` flag when the Alertmanager/Prometheus resource defines a non-nil `.spec.web` field.

Thanks for your review. Got it. Fixed.

File: .github/workflows/checks.yaml [hunk 4]
```diff
@@ -58,6 +58,7 @@ jobs:
         with:
           version: v1.55.2
           args: --timeout 10m0s
+          skip-pkg-cache: true
```
Comment (simonpasquier) [inline .github/workflows/checks.yaml:61]: it seems unrelated to the PR. Can you revert?

**Review by simonpasquier — COMMENTED** [review]
LGTM otherwise

File: .github/workflows/checks.yaml (hunk 4 above)
Comment (dongjiang1989) [reply .github/workflows/checks.yaml]: Thanks. Fixed it.

**Review by simonpasquier — APPROVED** [review]
Thanks!
